pkg/services/query/query.go

```diff
@@ -90,18 +90,41 @@ func (s *Service) QueryData(ctx context.Context, user *user.SignedInUser, skipCa
 	g, ctx := errgroup.WithContext(ctx)
 	results := make([]backend.Responses, len(byDataSource))
 
+	// Create panic recovery function for loop below
+	recoveryFn := func(results *[]backend.Responses, queries []*simplejson.Json) {
+		if r := recover(); r != nil {
+			var err error
+			s.log.Error("query datasource panic", "error", r, "stack", log.Stack(1))
+			if theErr, ok := r.(error); ok {
+				err = theErr
+			} else if theErrString, ok := r.(string); ok {
+				err = fmt.Errorf(theErrString)
+			} else {
+				err = fmt.Errorf("unexpected error, see the server log for details")
+			}
+			// Because of the panic, there is no valid response for any query. Append an error for each one.
+			*results = append(*results, buildErrorResponses(err, queries))
+		}
+	}
+
+	// Query each datasource concurrently
 	for _, queries := range byDataSource {
 		rawQueries := make([]*simplejson.Json, len(queries))
 		for i := 0; i < len(queries); i++ {
 			rawQueries[i] = queries[i].rawQuery
 		}
 		g.Go(func() error {
```

Comment:
Something else to consider is that this could create an unlimited amount of goroutines and execute them all in parallel. With a certain amount of queries from different datasources I think this could easily consume huge amount of CPU/RAM. Maybe you would like to limit the amount of parallelism allowed?

Some potential inspiration from the SQL datasources, even though they don't limit parallelism, but are using channels instead: https://github.com/grafana/grafana/blob/93c1fbbe3f3996e2e0d39d39acb235c802d37ed4/pkg/tsdb/sqleng/sql_engine.go#L176-L203
https://github.com/grafana/grafana/blob/93c1fbbe3f3996e2e0d39d39acb235c802d37ed4/pkg/tsdb/sqleng/sql_engine.go#L218-L230